proposed/2022/Additional-Metadata-Attributes.md

```diff
@@ -0,0 +1,76 @@
+# Additional Metadata Attributes
+
+- Author Name [@redth](https://github.com/redth)
+- Start Date (2022-10-190)
+- GitHub Issue (GitHub Issue link)
+- GitHub PR (GitHub PR link)
+
+## Summary
+
+Allow the inclusion of additional metadata properties in package authoring and allow them to be used in search queries.
+
+## Motivation 
+
+MAUI (and previously Xamarin) has an ecosystem of 'binding' library NuGet packages which surface native platform library usage within apps through C# API projections.  These native platform libraries are often available on their native package management system counterparts (ie: Maven, CocoaPods, etc.).  Typically binding nuget packages redistribute the native libraries within them.
+
+It's currently challenging/impossible to discern which NuGet packages map to a given native platform library package identity.  
+
+Moving the ecosystem forward, it would be very helpful to be able to programmatically determine the native platform library identities to:
+
+1. Deduplicate inclusion of packages through transitive dependencies (which often result in native toolchain build errors if not resolved)
+2. Discover eligible packages to fulfill dependencies of known native platform libraries (ie: integrating the output of a gradle build and automatically matching up / suggesting existing packages for dependencies)
+
+Identity can consist of multiple attributes, for example a Maven package has:
+
+- Group Id (eg: `com.company.product`)
+- Artifact Id (eg: `ProductSdk`)
```

Comment:
I'm not sure I fully understand the question here, could you elaborate?  It might be nice to support additional types and query operators if that's the basic question?  I guess if something is being considered, may as well consider all possible useful cases.

Reply:
Oh sorry, I'm not really _that_ familiar with how they built up their model.  

Having the ability to query on the version as a version number would be useful in an "AND" query scenario, I had just considered the MVP implementation of this not needing to do that since the result set matching only the `GroupId` and `ArtifactId` would presumably be small enough to iterate over versions of the results (also if it wasn't clear, different NuGet package _versions_ would potentially have different `attr_MavenVersion:1.2.3`  values).  The one potential gotcha here is that Maven's versioning rules might be a bit different than NuGet (though I _think_ semver is adopted there too). 

One other consideration though is the maven version might be used to assert if it satisfies a Maven version range - again, similar to NuGet's version range semantics, but not necessarily identical in rules/implementation.  For the binding helpers project/experiment I linked in the proposal, this is part of the process, so in this case the matching `GroupId/ArtifactId` results would still need to be iterated over, asserting each version's maven range compatibility.  Long way to say that there's maybe too many operators to consider for querying by version to make the effort of adding some simple >= particularly useful in value?  This is just one example though, and maybe that would be valuable for other scenarios.

Reply:
Sorry for coming late to the show. 

I maintain:

-   AndroidX (AX) 

    https://github.com/xamarin/AndroidX

-   GooglePlayServices, Firebase, MLKit (GPS-FB-MLKit)

    https://github.com/xamarin/GooglePlayServicesComponents

And when I get some air I work on "bindings improvements" which should improve productivity and more.

So, I will express my opinion only for Android (.NET for Android - formerly Xamarin.Android), though
IMO this should be extended to .NET for iOS and maybe other platforms.

"Bindings improvements" include

-   discoverability/identification (both Maven and NuGet)

-   dependency graphs (both Maven and NuGet)

    Here I work on porting my graph theory library from c++ to c# and thanks to Generic Math it is
    a lot easier nowdays.

    https://devblogs.microsoft.com/dotnet/dotnet-7-generic-math/
    https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/

    I need formal methods to build graph/trees, find leaves (all nodes) and detect dependency cycles
    for both Maven and NuGet.

I am already using some of the utilities in our repos for

-   building `cgmanifest.json`

    https://github.com/xamarin/AndroidX/blob/main/cgmanifest.json

    https://github.com/xamarin/GooglePlayServicesComponents/blob/main/cgmanifest.json

-   detecting dates when NuGet packages were published

Up to recently we have added Maven fully qualified metadata for artifact in 2 forms:

-   `artifact=androidx.compose.material:material-ripple` 
-   `artifact_versioned=androidx.compose.material:material-ripple:1.0.5`

to nuget fields 

- Description
- Summary (sometimes)
- Tags

Visible here:

https://api.nuget.org/v3/registration5-gz-semver2/xamarin.androidx.compose.material.ripple/index.json

This was OK and I am able to use server side NuGet protocol (HttpClient + JSON/XML parsing) to increase productivity of maintenance on both of repos.

Last updates .NET for Android team decided to keep this information only in `Tags` node.

-   identifying binaries used (either distributed or downloaded by the package during the build)

-   dependency identification

    -   type 
        -   maven
        -   native
    -   identity
    -   version

With this there would be 1:1 mapping from NuGet package (versioned) to Maven/Native package (versioned)
This would help maintainers with 

1.  keeping track of published (bound Maven or native libraries),

    Getting data for latest nuget package and mapping it to maven fully qualified versioned id
    would ease discoverability what is to be updated.

2.  updates and 

    see 1.

3.  troubleshooting 

    Primarily checking dependency graphs 

    -   for duplicate transitive dependencies (possibly with different versions)

4.  security checks (component governance)

5.  curation (currated package publishing)

    With lowering the bar for bindings via "bindings improvements" it is to be expected to have 
    flood of bindings packages.
    NuGet publishing proces could add step to verify if given Maven/Native artifact is already 
    published in some other NuGet package.



*   Maven

    *   project

        ```xml
        <ItemGroup>
            <PackageAttribute Include="maven.GroupId" Value="androidx.activity" />
            <PackageAttribute Include="maven.ArtifactId" Value="activity" />
            <PackageAttribute Include="maven.VersionId" Value="1.6.0" />
        </ItemGroup>
        ```

        NOTE: this could be derived from curernt (and future) .NET for Android (Xamarin.Android) 
        BuildActions for binding artifacts (`Embedd`)

    *   nuspec

        ```xml
        <!--
            ... snip
        -->
        <package>
            <metadata>
                <attributes>
                <attribute key="maven.GroupId">androidx.activity</attribute>
                <attribute key="maven.ArtifactId">activity</attribute>
                <attribute key="maven.Version">1.6.0</attribute>
                </attributes>
            </metadata>
        <!--
            ... snip
        -->
        </package>
        ```   

*   Native

    *   project (packaging)

        ```xml
        <ItemGroup>
            <AndroidNativeLibrary Include="path/to/libfoo.so">
                <Abi>armeabi</Abi>
            </AndroidNativeLibrary>
        </ItemGroup>
        ```

    *   nuspec

        ```xml
        <package>
            <metadata>
                <attributes>
                <attribute key="native.LibraryName">libfoo</attribute>
                <attribute key="native.Version">1.6.0</attribute>
                </attributes>
            </metadata>
        </package>
        ```